primitives/arkworks/src/bls12_377.rs

```diff
@@ -0,0 +1,198 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Hashing Functions.
+
+#![warn(missing_docs)]
+
+use crate::utils::serialize_result;
+use ark_bls12_377::{g1, g2, Bls12_377, G1Affine, G1Projective, G2Affine, G2Projective};
+use ark_ec::{
+	models::CurveConfig,
+	pairing::{MillerLoopOutput, Pairing},
+	short_weierstrass::SWCurveConfig,
+};
+use ark_serialize::{CanonicalDeserialize, Compress, Validate};
+use ark_std::io::Cursor;
+use sp_std::vec::Vec;
+
+/// Compute multi miller loop through arkworks
+pub fn multi_miller_loop(a_vec: Vec<Vec<u8>>, b_vec: Vec<Vec<u8>>) -> Vec<u8> {
+	let g1: Vec<_> = a_vec
+		.iter()
+		.map(|a| {
+			let cursor = Cursor::new(a);
+			<Bls12_377 as Pairing>::G1Affine::deserialize_with_mode(
+				cursor,
+				Compress::Yes,
+				Validate::No,
+			)
+			.map(<Bls12_377 as Pairing>::G1Prepared::from)
+			.unwrap()
+		})
+		.collect();
+	let g2: Vec<_> = b_vec
+		.iter()
+		.map(|b| {
+			let cursor = Cursor::new(b);
+			<Bls12_377 as Pairing>::G2Affine::deserialize_with_mode(
+				cursor,
+				Compress::Yes,
```

Comment:
I doubt you want to compress curve points, maybe G2 points would be better partially compressed or something, but not necessarily.  It's possible this causes some of your odd benchmarks.

Reply:
Point compression is now disabled.